### ISSUE TITLE
Strange behavior for `nothing` initialized at top level  

### ISSUE THREAD

**Issue by sgalles** [ceylon/ceylon-compiler#2034]
This seems inconsistent :
```ceylon
variable Anything a = nothing;
shared void run(){
    a = 42;    // AssertionError at runtime here.
    print(a);
}
```
and

```ceylon
variable Anything a = nothing;
shared void run(){
    a = 42;    // *No* AssertionError at runtime
}
```
FTR : The JS backend prints 42 for the first example...

**Comment by lucaswerkmeister** [comment]
Generated code:
```java
package tmp;

public final class run_ {
    
    private run_() {
    }
    
    public static void run() {
        .tmp.a_.set_(.ceylon.language.Integer.instance(42L));
        .ceylon.language.print_.print(.tmp.a_.get_());
    }
    
    public static void main(.java.lang.String[] args) {
        .ceylon.language.process_.get_().setupArguments(args);
        .tmp.run_.run();
    }
}
final class a_ {
    
    private a_() {
    }
    private static .java.lang.Object value;
    private static volatile boolean $init$value = false;
    private static final .java.lang.Throwable $initException$;
    static {
        try {
            value = .ceylon.language.nothing_.get_();
            $initException$ = null;
            $init$value = true;
        } catch (.java.lang.Throwable x) {
            $initException$ = x;
            value = null;
            $init$value = false;
        }
    }
    
    public static .java.lang.Object get_() {
        if (a_.$init$value) {
            return value;
        } else {
            if ($initException$ != null) .com.redhat.ceylon.compiler.java.Util.rethrow($initException$);
            throw new .ceylon.language.InitializationError("Cyclic initialization trying to read the value of \'a\' before it was set");
        }
    }
    
    public static void set_(final .java.lang.Object a) {
        value = a;
    }
}
```
What’s happening:

1. `a_.set_(42)` causes the static initializer of `a_` to be run.
2. Within the static initializer, `nothing_.get_()` is run and throws an `AssertionError`.
3. The static initializer catches any `Throwable`, including that error, and stores it in `$initException$`. Just to be safe, it also resets the `$init$value` flag.
4. The `set_` call itself runs, setting `a_.value` but leaving the `$init$value` flag unmodified.
5. In `run()`, line 4, we try to `get_()` the value of `a`.
6. `get_()` sees that `$init$value` is false (i. e. the value is uninitialized), and throws the exception that was stored by the static initializer.

If this is incorrect (you *could* argue that `a` is indeed not yet initialized… but that’s inconsistent with JS, and it’s probably useful to be able to circumvent a toplevel’s initial value), then `set_()` should probably set the `$init$value` flag.

@sgalles: The stacktrace of the exception indicates that line 3 (`a = 42`) caused the exception, but it only leaves `a` in line 4 (`print(a)`). If you insert another `print` (or anything else) in between them, it’s still executed.

**Comment by lucaswerkmeister** [comment]
> then set_() should probably set the $init$value flag.

Actually… with that solution the initial value is still evaluated, just not used. If the initializer expression has side effects, I’m not sure if that’s correct or not.

OTOH, I’m not sure if it’s even possible to avoid that. Probably not without ditching the static initializer, since that gets run before `set_()` could possibly inform it that it’s not gonna be needed.

**Comment by lucaswerkmeister** [comment]
> I’m not sure if that’s correct or not

Well there’s a simpler metric than ethics: Backend comparison! :D
```ceylon
T printIt<T>(T t) { print(t); return t; }
variable Anything a = printIt("init");
shared void run() {
    a = "value";
    print(a);
}
```
Java:
> init
> value

JS:
> value

So that’s also inconsistent at the moment.

**Comment by FroMage** [comment]
Well, it's not clear at all to me that the Java behaviour is the wrong one. It depends on what the spec would say, but it seems very intuitive to me that given eager evaluation with `=` then the attribute _must_ be defined and initialised before it is set to a new value. Otherwise the initialisation of the variable is entirely skipped and that doesn't seem right.

**Comment by FroMage** [comment]
Of course, it would be different with `=>` and lazy evaluation, but with `=` I'm inclined to say the spec should bless the JVM backend behaviour and close this issue.

**Comment by gavinking** [comment]
@FroMage well the issue is the difference between the two code examples right at the top of this issue.

**Comment by FroMage** [comment]
Well, so we have to define what needs to happen then. IMO initialisation should happen before we do the `set`, though in the case of deferred initialisation, that happens to be right where `set` is called, so `set` should throw in both cases. From that, it follows that it throws before the new value is set, and therefore every call to read and write should throw forever. Agreed?

**Comment by gavinking** [comment]
Sounds good to me.

**Comment by FroMage** [comment]
So, what is confusing is that the exception which is thrown is saved from the place where we evaluated it, and so the line numbers point to where the initialisation happened, and not when we tried every subsequent evaluation. But I don't think I can change that.

**Comment by FroMage** [comment]
So this is fixed and will throw on every get/set. I suppose the spec needs to bless this and the JS backend too.

**Comment by FroMage** [comment]
Done, and opened a JS issue for it.
